src/plans/keywords.jl

```diff
@@ -0,0 +1,252 @@
+"""
+    Keywords
+
+A small internal struct to represent a set of keywords,
+
+# Fields
+
+* `accepted=Set{Symbol}()` a `Set` of symbols of keywords a certain function accepts
+* `deprecated=Set{Symbol}()` a `Set` of symbols of keywords a certain function has deprecated
+* `from=nothing` the function the keywords are (directly or indirectly) come from or accepted in.
+  to indicate that these are not associated with a certain function, use `nothing`.
+  to Indicate ane empty set, use `nothing`.
+* `origins` a dictionary that specifies for every keyword the function it is passed to.
+  this usually should point to the function it is _directly_ passed to.
+
+# Constructor
+
+    Keywords(
+        accepted=Set{Symbol}(), deprecated=Set{Symbol}();
+        from::Type=nothing)
+
+Generate a Keywords wrapper, where both default to being the empty set.
+For pretty printing you can provide a type they belong to.
+"""
+struct Keywords{I}
+    accepted::Set{Symbol}
+    deprecated::Set{Symbol}
+    origins::Dict{Symbol, Union{T, Vector{T}} where {T <: Union{Function, Type}}}
+end
+function Keywords(
+        accepted::Set{Symbol} = Set{Symbol}(),
+        deprecated::Set{Symbol} = Set{Symbol}();
+        from = nothing,
+        origins = nothing,
+    )
+    if !isnothing(from)
+        _origins = isnothing(origins) ? Dict{Symbol, Any}() : origins
```

Comment:
`_origins` has three different types here: `Dict{Symbol, Union{T, Vector{T}} where {T <: Union{Function, Type}}}`, `Dict{Symbol, Any}` or `_origins = Dict{Symbol, Union{<:Function, <:Type}}`. It'd be preferable to stick to one. I personally prefer `Any` over `Union{Function, Type}` because `Any` would also work for functors.

BTW, is a `Vector` value in `origins` a remnant from older drafts or something you'd like to support? Why not having it always be a vector for consistency?

Reply:
Ok, sure, this is also not performance critical since it is meant to only be called outside of those.

We could at least do `Dict{Symbol, Vector{V}}` and make the `V` a parameter in the corresponding struct?

Reply:
Hm that is not so easy, every now and then I need it to be nothing as well

Reply:
Making this faster than what what it would be with `Dict{Symbol, Vector{Any}}` would be challenging. Having such `V` would be a part of it, sure, but you'd also have to set `V` to small unions of concrete types and inspect that union splitting actually can happen, which is IMO not worth the small benefits it could provide.